File: app/src/main/java/com/aaron/yespdf/settings/SettingsAdapter.java
package com.aaron.yespdf.settings;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.SeekBar;
import android.widget.Spinner;
import android.widget.Switch;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.aaron.yespdf.R;
import com.aaron.yespdf.common.App;
import com.aaron.yespdf.common.Settings;
import com.aaron.yespdf.common.UiManager;
import com.aaron.yespdf.common.event.MaxRecentEvent;
import com.blankj.utilcode.util.ConvertUtils;

import org.greenrobot.eventbus.EventBus;

import java.util.Arrays;
import java.util.List;

/**
 * @author Aaron aaronzzxup@gmail.com
 */
class SettingsAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    private static final int ITEM_COUNT = 5;

    private static final int TYPE_SWITCH = 0;
    private static final int TYPE_SEEKBAR = 1;
    private static final int TYPE_NUM_PICKER = 2;

    private static final int POS_COLOR_REVERSE = 0;
    private static final int POS_VOLUME_CONTROL = 1;
    private static final int POS_SHOW_STATUS_BAR = 2;
    private static final int POS_NUM_PICKER = 3;
    private static final int POS_SCROLL_LEVEL = 4;

    private MaxRecentEvent mMaxRecentEvent;
    private List<String> mMaxRecentCounts;

    SettingsAdapter() {
        mMaxRecentEvent = new MaxRecentEvent();
        String[] array = App.getContext().getResources().getStringArray(R.array.max_recent_count);
        mMaxRecentCounts = Arrays.asList(array);
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        Context context = parent.getContext();
        LayoutInflater inflater = LayoutInflater.from(context);
        if (viewType == TYPE_SWITCH) {
            View itemView = inflater.inflate(R.layout.app_recycler_item_settings_switch, parent, false);
            SwitchHolder holder = new SwitchHolder(itemView);
            holder.itemView.setOnClickListener(v -> {
                int pos = holder.getAdapterPosition();
                holder.switcher.setChecked(!holder.switcher.isChecked());
                tapOptions(holder.switcher, pos);
            });
            return holder;
        } else if (viewType == TYPE_NUM_PICKER) {
            View itemView = inflater.inflate(R.layout.app_recycler_item_settings_recent_count, parent, false);
            MaxRecentHolder holder = new MaxRecentHolder(itemView);
            holder.itemView.setOnClickListener(v -> holder.spinner.performClick());
            holder.spinner.setDropDownHorizontalOffset(-ConvertUtils.dp2px(46));
            holder.spinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
                @Override
                public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                    String maxRecentCount = mMaxRecentCounts.get(position);
                    holder.tvCount.setText(maxRecentCount);
                    Settings.setMaxRecentCount(maxRecentCount);
                    EventBus.getDefault().post(mMaxRecentEvent);
                }

                @Override
                public void onNothingSelected(AdapterView<?> parent) {
                }
            });
            return holder;
        }
        View itemView = inflater.inflate(R.layout.app_recycler_item_settings_seekbar, parent, false);
        SeekbarHolder holder = new SeekbarHolder(itemView);
        holder.seekBar.setMax(9);
        holder.seekBar.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {

            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {

            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {
                int level = seekBar.getProgress() + 1;
                UiManager.showShort(context.getString(R.string.app_cur_level_is) + level);
                Settings.setScrollLevel(level);
            }
        });
        return holder;
    }

    private void tapOptions(Switch switcher, int pos) {
        switch (pos) {
            case POS_COLOR_REVERSE:
                Settings.setNightMode(switcher.isChecked());
                break;
            case POS_VOLUME_CONTROL:
                Settings.setVolumeControl(switcher.isChecked());
                break;
            case POS_SHOW_STATUS_BAR:
                Settings.setShowStatusBar(switcher.isChecked());
                break;
        }
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder viewHolder, int position) {
        if (viewHolder instanceof SwitchHolder) {
            SwitchHolder holder = (SwitchHolder) viewHolder;
            switch (position) {
                case POS_COLOR_REVERSE:
                    holder.tvTitle.setText(R.string.app_color_reverse);
                    holder.switcher.setChecked(Settings.isNightMode());
                    break;
                case POS_VOLUME_CONTROL:
                    holder.tvTitle.setText(R.string.app_volume_control);
                    holder.switcher.setChecked(Settings.isVolumeControl());
                    break;
                case POS_SHOW_STATUS_BAR:
                    holder.tvTitle.setText(R.string.app_show_status_bar);
                    holder.switcher.setChecked(Settings.isShowStatusBar());
                    break;
            }
        } else if (viewHolder instanceof SeekbarHolder) {
            SeekbarHolder holder = (SeekbarHolder) viewHolder;
            holder.tvTitle.setText(R.string.app_scroll_velocity);
            holder.seekBar.setProgress((int) Settings.getScrollLevel() - 1);
        } else if (viewHolder instanceof MaxRecentHolder) {
            MaxRecentHolder holder = (MaxRecentHolder) viewHolder;
            holder.tvTitle.setText(R.string.app_max_recent_count);
            holder.tvCount.setText(Settings.getMaxRecentCount());
            holder.spinner.setSelection(mMaxRecentCounts.indexOf(Settings.getMaxRecentCount()));
        }
    }

    @Override
    public int getItemViewType(int position) {
        if (position == POS_SCROLL_LEVEL) {
            return TYPE_SEEKBAR;
        } else if (position == POS_NUM_PICKER) {
            return TYPE_NUM_PICKER;
        }
        return TYPE_SWITCH;
    }

    @Override
    public int getItemCount() {
        return ITEM_COUNT;
    }

    static class SwitchHolder extends RecyclerView.ViewHolder {
        TextView tvTitle;
        Switch switcher;

        SwitchHolder(@NonNull View itemView) {
            super(itemView);
            tvTitle = itemView.findViewById(R.id.app_tv_title);
            switcher = itemView.findViewById(R.id.app_switch);
        }
    }

    static class SeekbarHolder extends RecyclerView.ViewHolder {
        TextView tvTitle;
        SeekBar seekBar;

        SeekbarHolder(@NonNull View itemView) {
            super(itemView);
            tvTitle = itemView.findViewById(R.id.app_tv_title);
            seekBar = itemView.findViewById(R.id.app_sb_scroll_level);
        }
    }

    static class MaxRecentHolder extends RecyclerView.ViewHolder {
        TextView tvTitle;
        TextView tvCount;
        Spinner spinner;

        MaxRecentHolder(@NonNull View itemView) {
            super(itemView);
            tvTitle = itemView.findViewById(R.id.app_tv_title);
            tvCount = itemView.findViewById(R.id.app_tv_count);
            spinner = itemView.findViewById(R.id.app_spinner);
        }
    }
}

File: app/src/main/java/com/aaron/yespdf/settings/SettingsAdapter.kt
package com.aaron.yespdf.settings

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.ViewGroup.MarginLayoutParams
import android.widget.AdapterView
import android.widget.AdapterView.OnItemSelectedListener
import android.widget.SeekBar
import android.widget.SeekBar.OnSeekBarChangeListener
import android.widget.Switch
import androidx.recyclerview.widget.RecyclerView
import com.aaron.yespdf.R
import com.aaron.yespdf.common.App
import com.aaron.yespdf.common.Settings
import com.aaron.yespdf.common.event.MaxRecentEvent
import com.blankj.utilcode.util.ConvertUtils
import kotlinx.android.synthetic.main.app_recycler_item_settings_recent_count.view.*
import kotlinx.android.synthetic.main.app_recycler_item_settings_seekbar.view.*
import kotlinx.android.synthetic.main.app_recycler_item_settings_switch.view.*
import kotlinx.android.synthetic.main.app_recycler_item_settings_switch.view.app_tv_title
import org.greenrobot.eventbus.EventBus

/**
 * @author Aaron aaronzzxup@gmail.com
 */
internal class SettingsAdapter : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    private val maxRecentEvent: MaxRecentEvent = MaxRecentEvent()
    private val maxRecentCounts: List<String>

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val context = parent.context
        val inflater = LayoutInflater.from(context)
        if (viewType == TYPE_SWITCH) {
            val itemView = inflater.inflate(R.layout.app_recycler_item_settings_switch, parent, false)
            val holder = SwitchHolder(itemView)
            holder.itemView.setOnClickListener {
                val pos = holder.adapterPosition
                holder.itemView.app_switch.isChecked = !holder.itemView.app_switch.isChecked
                tapOptions(holder.itemView.app_switch, pos)
            }
            return holder
        } else if (viewType == TYPE_NUM_PICKER) {
            val itemView = inflater.inflate(R.layout.app_recycler_item_settings_recent_count, parent, false)
            val holder = MaxRecentHolder(itemView)
            holder.itemView.setOnClickListener { holder.itemView.app_spinner.performClick() }
            holder.itemView.app_spinner.dropDownHorizontalOffset = -ConvertUtils.dp2px(46f)
            holder.itemView.app_spinner.onItemSelectedListener = object : OnItemSelectedListener {
                override fun onItemSelected(parent: AdapterView<*>?, view: View, position: Int, id: Long) {
                    val maxRecentCount = maxRecentCounts[position]
                    holder.itemView.app_tv_count.text = maxRecentCount
                    Settings.maxRecentCount = maxRecentCount
                    EventBus.getDefault().post(maxRecentEvent)
                }

                override fun onNothingSelected(parent: AdapterView<*>?) {}
            }
            return holder
        }
        val itemView = inflater.inflate(R.layout.app_recycler_item_settings_seekbar, parent, false)
        val holder = SeekbarHolder(itemView)
        holder.itemView.app_sb_scroll_level.max = 14
        holder.itemView.app_sb_scroll_level.setOnSeekBarChangeListener(object : OnSeekBarChangeListener {
            override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {}
            override fun onStartTrackingTouch(seekBar: SeekBar) {}
            override fun onStopTrackingTouch(seekBar: SeekBar) {
                val level = seekBar.progress + 1
//                UiManager.showShort(context.getString(R.string.app_cur_level, level))
                Settings.scrollLevel = level.toLong()
            }
        })
        return holder
    }

    private fun tapOptions(switcher: Switch, pos: Int) {
        when (pos) {
            POS_VOLUME_CONTROL -> Settings.volumeControl = switcher.isChecked
            POS_CLICK_FLIP_PAGE -> Settings.clickFlipPage = switcher.isChecked
            POS_SHOW_STATUS_BAR -> Settings.showStatusBar = switcher.isChecked
            POS_KEEP_SCREEN_ON -> Settings.keepScreenOn = switcher.isChecked
            POS_LINEAR_LAYOUT -> Settings.linearLayout = switcher.isChecked
            POS_SCROLL_SHORTCUT -> Settings.scrollShortCut = switcher.isChecked
        }
    }

    override fun onBindViewHolder(viewHolder: RecyclerView.ViewHolder, position: Int) {
        when (viewHolder) {
            is SwitchHolder -> {
                when (position) {
                    POS_VOLUME_CONTROL -> {
                        viewHolder.itemView.app_tv_title.setText(R.string.app_volume_control)
                        viewHolder.itemView.app_switch.isChecked = Settings.volumeControl
                    }
                    POS_CLICK_FLIP_PAGE -> {
                        viewHolder.itemView.app_tv_title.setText(R.string.app_click_flip_page)
                        viewHolder.itemView.app_switch.isChecked = Settings.clickFlipPage
                    }
                    POS_SHOW_STATUS_BAR -> {
                        viewHolder.itemView.app_tv_title.setText(R.string.app_show_status_bar)
                        viewHolder.itemView.app_switch.isChecked = Settings.showStatusBar
                    }
                    POS_KEEP_SCREEN_ON -> {
                        viewHolder.itemView.app_tv_title.setText(R.string.app_keep_screen_on)
                        viewHolder.itemView.app_switch.isChecked = Settings.keepScreenOn
                    }
                    POS_LINEAR_LAYOUT -> {
                        viewHolder.itemView.app_tv_title.setText(R.string.app_linear_layout)
                        viewHolder.itemView.app_switch.isChecked = Settings.linearLayout
                    }
                    POS_SCROLL_SHORTCUT -> {
                        viewHolder.itemView.app_tv_title.setText(R.string.app_scroll_shortcut)
                        viewHolder.itemView.app_switch.isChecked = Settings.scrollShortCut
                    }
                }
            }
            is SeekbarHolder -> {
                viewHolder.itemView.app_tv_title.setText(R.string.app_scroll_velocity)
                viewHolder.itemView.app_sb_scroll_level.progress = Settings.scrollLevel.toInt() - 1
            }
            is MaxRecentHolder -> {
                viewHolder.itemView.app_tv_title.setText(R.string.app_max_recent_count)
                viewHolder.itemView.app_tv_count.text = Settings.maxRecentCount
                viewHolder.itemView.app_spinner.setSelection(maxRecentCounts.indexOf(Settings.maxRecentCount))
            }
        }
    }

    override fun getItemViewType(position: Int): Int {
        if (position == POS_SCROLL_LEVEL) {
            return TYPE_SEEKBAR
        } else if (position == POS_NUM_PICKER) {
            return TYPE_NUM_PICKER
        }
        return TYPE_SWITCH
    }

    override fun getItemCount(): Int {
        return ITEM_COUNT
    }

    internal class SwitchHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        init {
            val mlp = itemView.layoutParams as MarginLayoutParams
            mlp.bottomMargin = ConvertUtils.dp2px(0.6f)
            itemView.layoutParams = mlp
        }
    }

    internal class SeekbarHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        init {
            val mlp = itemView.layoutParams as MarginLayoutParams
            mlp.topMargin = ConvertUtils.dp2px(0.6f)
            itemView.layoutParams = mlp
        }
    }

    internal class MaxRecentHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        init {
            val mlp = itemView.layoutParams as MarginLayoutParams
            mlp.topMargin = ConvertUtils.dp2px(12f)
            itemView.layoutParams = mlp
        }
    }

    companion object {
        private const val ITEM_COUNT = 8

        private const val TYPE_SWITCH = 0
        private const val TYPE_SEEKBAR = 1
        private const val TYPE_NUM_PICKER = 2

        private const val POS_VOLUME_CONTROL = 0
        private const val POS_CLICK_FLIP_PAGE = 1
        private const val POS_SHOW_STATUS_BAR = 2
        private const val POS_KEEP_SCREEN_ON = 3
        private const val POS_LINEAR_LAYOUT = 4
        private const val POS_SCROLL_SHORTCUT = 5
        private const val POS_NUM_PICKER = 6
        private const val POS_SCROLL_LEVEL = 7
    }

    init {
        val array = App.getContext().resources.getStringArray(R.array.max_recent_count)
        maxRecentCounts = listOf(*array)
    }
}